### PR TITLE
feat(webhooks): DOMA-4409 Webhook models added

### PR DESCRIPTION
**Pull request by SavelevMatthew**

### CODE REVIEW

File: apps/condo/domains/user/schema/User.js [hunk 1]
```diff
@@ -205,7 +207,15 @@ const User = new GQLListSchema('User', {
...
     },
-    plugins: [uuided(), versioned(), tracked(), softDeleted(), dvAndSender(), historical()],
+    plugins: [
+        uuided(),
+        versioned(),
+        tracked(),
+        softDeleted(),
+        dvAndSender(),
+        historical(),
+        webHooked(webhookModelValidator),
```
Comment (SavelevMatthew) [inline apps/condo/domains/user/schema/User.js:217]: Is `OK` to call it like that? Or may be add wrapper to combine validator with plugin?

File: apps/condo/webhooks/webhooks.test.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
+const { makeLoggedInAdminClient, makeClient } = require('@condo/keystone/test.utils')
+const { makeClientWithSupportUser } = require('@condo/domains/user/utils/testSchema')
+const { makeClientWithNewRegisteredAndLoggedInUser } = require('@condo/domains/user/utils/testSchema')
+const { WebhookTests } = require('@condo/webhooks/schema/models/Webhook.test')
+const { WebhookSubscriptionTests } = require('@condo/webhooks/schema/models/WebhookSubscription.test')
+
+async function initializeActors () {
```
Comment (SavelevMatthew) [inline apps/condo/webhooks/webhooks.test.js:7]: Storing it here is `OK`?

File: packages/webhooks/plugins/webHooked.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+const { GQL_SCHEMA_PLUGIN } = require('@condo/keystone/plugins/utils/typing')
+const { composeNonResolveInputHook } = require('@condo/keystone/plugins/utils')
+
+const plugin = (fn) => {
```
Comment (SavelevMatthew) [inline packages/webhooks/plugins/webHooked.js:4]: Custom wrapper, since default one call hook on `Test` schema, and then it predictably throw Error

**Comment by akarjakin** [comment]
Would these webhooks be executed in resolvers or in afterChange?

File: apps/condo/webhooks/webhooks.test.js (hunk 2 above)
Comment (SavelevMatthew) [reply apps/condo/webhooks/webhooks.test.js]: Moved tests to `common` domain

Comment (SavelevMatthew) [reply apps/condo/webhooks/webhooks.test.js]: validator is still under discussion

File: apps/condo/domains/user/schema/User.js (hunk 1 above)
Comment (pahaz) [inline apps/condo/domains/user/schema/User.js:217]: it's better to define `webhookModelValidator` inside the plugin logic

File: packages/webhooks/plugins/webHooked.js (hunk 3 above)
Comment (pahaz) [reply packages/webhooks/plugins/webHooked.js:6]: use `const validator  = require('@condo/webhooks/model-validator')` and the use `validator. CURRENT_VALIDATOR` or create a function API for that

Comment (SavelevMatthew) [reply packages/webhooks/plugins/webHooked.js:5]: https://github.com/open-condo-software/condo/pull/2497/commits/86e9c28e85f95fab65325596b6d4e74903d0d7c9

File: apps/condo/webhooks/webhooks.test.js (hunk 2 above)
Comment (SavelevMatthew) [reply apps/condo/webhooks/webhooks.test.js]: changed logic of initialising validator in: https://github.com/open-condo-software/condo/pull/2497/commits/86e9c28e85f95fab65325596b6d4e74903d0d7c9

**Review by dkoviazin — APPROVED** [review]
How the security will be made on remote system? IP restriction?
Some systems allows to store a key (to sign requests) in web hook settings or are using basic auth

File: apps/condo/domains/user/schema/User.js (hunk 1 above)
Comment (SavelevMatthew) [reply apps/condo/domains/user/schema/User.js]: OUTDATED

Comment (SavelevMatthew) [reply apps/condo/domains/user/schema/User.js]: Outdated, changed it in: https://github.com/open-condo-software/condo/pull/2497/commits/86e9c28e85f95fab65325596b6d4e74903d0d7c9

**Comment by SavelevMatthew** [comment]
> How the security will be made on remote system? IP restriction?
> Some systems allows to store a key (to sign requests) in web hook settings or are using basic auth

For MVP will send data "as is". Then I'm planning to add some `SECRET` field to sign data

File: apps/condo/jest.config.js [hunk 4]
```diff
@@ -3,14 +3,23 @@ module.exports = {
         {
             displayName: 'schema',
             testEnvironment: 'node',
-            testMatch: [`${__dirname}/schema/**/*.test.js`, `${__dirname}/domains/**/schema/*.test.js`],
+            testMatch: [
+                `${__dirname}/schema/**/*.test.js`,
```
Comment (pahaz) [inline apps/condo/jest.config.js:7]: It’s inside the common. Why we need it?

File: packages/webhooks/model-validator/validator.js [hunk 5]
```diff
@@ -14,26 +14,22 @@ class WebHookModelValidator {
...
-     * Initialize validator by loading and converting GQL schema
-     * @returns {Promise<void>}
-     */
-    async init () {
-        const gqlSchema = await loadSchema(this.schemaPath, {
+        const gqlSchema = loadSchemaSync(this.schemaPath, {
             loaders: [new GraphQLFileLoader()],
         })
-        const gqlIntrospection = await introspectionFromSchema(gqlSchema)
+        const gqlIntrospection = introspectionFromSchema(gqlSchema)
```
Comment (pahaz) [inline packages/webhooks/model-validator/validator.js:32]: It looks like every model should have big object inside the memory or not?

Comment (SavelevMatthew) [reply packages/webhooks/model-validator/validator.js:32]: They have compiled AJV, which are built from minified condo shema

File: packages/webhooks/schema/models/WebhookSubscription.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            webhook: {
+                schemaDoc: 'Link to a webhook containing information about integration',
+                type: Relationship,
+                ref: 'Webhook',
+                isRequired: true,
+                knexOptions: { isNotNullable: true },
+                kmigratorOptions: { null: false, on_delete: 'models.CASCADE' },
+            },
+            url: {
+                schemaDoc: 'Webhook target URL to which requests will be sent. Overrides url from webhook relation. ' +
```
Comment (pahaz) [inline packages/webhooks/schema/models/WebhookSubscription.js:61]: Why no just to create another webhook?

File: packages/webhooks/schema/models/WebhookSubscription.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+                        return WebHookModelValidator.normalizeFieldsString(resolvedData[fieldPath])
+                    },
+                    validateInput: validateFields,
+                },
+            },
+            filters: {
+                schemaDoc: 'Filters which is stored in JSON and used to filter models sent by the webhook. ' +
+                    'Examples of filters can be found in ModelWhereInput GQL type, ' +
+                    'where Model is name of your model',
+                type: Json,
```
Comment (pahaz) [inline packages/webhooks/schema/models/WebhookSubscription.js:102]: It’s a little confusing. Why fields is a strung, but the filters field is Json

**Review by pahaz — APPROVED** [review]

File: apps/condo/jest.config.js (hunk 4 above)
Comment (SavelevMatthew) [reply apps/condo/jest.config.js:7]: I've just reformatted code....Nothing was added 😄

File: packages/webhooks/model-validator/validator.js (hunk 5 above)
Comment (SavelevMatthew) [reply packages/webhooks/model-validator/validator.js:32]: As far as I know, Ajv has their own uri-system for preventing schema from being duplicated

Comment (SavelevMatthew) [reply packages/webhooks/model-validator/validator.js:32]: image

File: packages/webhooks/schema/models/WebhookSubscription.js (hunk 6 above)
Comment (SavelevMatthew) [reply packages/webhooks/schema/models/WebhookSubscription.js:61]: Webhook will contain additional info about integration, such as secrets for encryption and etc

**Pull request merged**